### PR TITLE
Add the oblique-shock 2D mesh for the multidimensional Euler solver

### LINKED ISSUE #404: Port the 2/3D Euler solver from the legacy solvcon code base
There are a couple of multi-dimensional Euler solvers implemented in https://github.com/solvcon/solvcon .  This issue tracks the progress to port them to solvcon.

Tasks:

- [x] Phase 1 -- Oblique-shock 2D mesh and boundary tagging
  - PR #885.
- [x] Phase 2 -- Python driver (new solvcon/multidim package)
  - PR #886.
- [ ] Phase 3 -- Visualization boundaries and solutions
  - PR #895.
  - PR #923.
- [ ] Phase 4 -- Correct the computing domain for the oblique shock reflection
- [ ] Phase 5 -- Run to steady state and validate

## Phase 1 -- Oblique-shock 2D mesh and boundary tagging

Build the 2D domain for the classic oblique shock reflection: a uniform supersonic inflow over a rectangular domain whose bottom is a slip wall carrying a wedge or inclined to deflect the flow by a fixed angle, producing an oblique shock that reflects off the opposite slip wall. Either generate the mesh programmatically (set `ndcrd`, `cltpn`, `clnds`, then `build_interior(do_metric=True)`, `build_boundary()`, `build_ghost()`, exactly as the existing tests and `_build_quad_channel` do) or add a dedicated `tests/data/oblique_shock.msh`. Because `add_bc` consumes face indices, provide a helper that returns the boundary face-index lists per physical role by geometric location: identify boundary faces via `fccls(ifc, 1) < 0`, then classify each by its `fccnd` position and `fcnml` direction into inlet (`x == xmin`), slip wall (the deflecting bottom and the reflecting top or wedge), and non-reflective outflow (`x == xmax`). Unit-test the classification: the inlet/wall/outflow sets must be non-empty, disjoint, and together cover every boundary face.

Results (PR #885): The mesh is generated programmatically; no `tests/data/oblique_shock.msh` was needed. `ObliqueShockMesher` in `solvcon/multidim/euler/oblique.py` builds the rectangular domain (default 3.0 x 1.0, 48 x 16) as structured quads, structured triangles, or jittered unstructured triangles (Delaunay via `_triangulate`), following the `build_interior(do_metric=True)` / `build_boundary()` / `build_ghost()` sequence. `classify_boundary(mh, tol)` identifies boundary faces via `fccls(ifc, 1) < 0` and classifies each by `fccnd` position and `fcnml` direction into the inlet, slip-wall, and non-reflective-outflow face-index lists that `add_bc` consumes. `tests/test_multidim_euler_oblique.py` verifies the meshes and that the three sets are non-empty, disjoint, and together cover every boundary face. Beyond the original scope, the PR also landed the `solvcon/multidim/` package scaffolding (`solvcon/multidim/__init__.py`, `solvcon/multidim/euler/__init__.py`, `solvcon/__init__.py`, the `setup.py` packages list) originally slated for phase 2, and a pilot feature (`solvcon/pilot/_oblique.py`, wired into `solvcon/pilot/_gui.py`) that draws the quad, triangle, and unstructured meshes from a menu. Note: the merged mesher keeps the bottom wall flat (no wedge), so the flow deflection is to be imposed by the inlet free-stream angle.

## Phase 2 -- Python driver (new solvcon/multidim package)

Add an `ObliqueShock` driver class to the `solvcon/multidim/` package.  Only make sure the solver runs through without crashing.

## Phase 3 -- Visualization boundaries and solutions

Add visualization UI for boundaries and solutions.  Add UI to start, pause, resume, and change parameters of the solver.

## Phase 4 -- Correct the computing domain for the oblique shock reflection

The oblique shock reflection should use a rectangular domain without wedge.

## Phase 5 -- Run to steady state and validate

Set up the oblique-shock reflection problem and run it to steady state.  Measure the oblique shock angle to make sure the solution is correct.

----

~~The porting will use the branch [`multidim`](https://github.com/solvcon/solvcon/tree/multidim).~~  Some work of the porting used the [`multidim`](https://github.com/solvcon/solvcon/tree/multidim) branch, but I found it makes tracking related issues harder.  I will go back to the [`master`](https://github.com/solvcon/solvcon/tree/master) branch.

----

- [x] #403 Adds the placeholder for porting code from the old solvcon repository.
- [x] #422 Add a simple 2D mesh file for testing.
- ~~[ ] Merge back to the [`master`](https://github.com/solvcon/solvcon/tree/master) branch.~~
- [ ] Port code to set up oblique shock wave problem in 2D.

### PR DESCRIPTION
**Pull request by yungyuc**
Set up the mesh for the classic oblique-shock reflection: a supersonic inflow over a rectangular domain whose bottom slip wall ramps into a wedge, with the reflecting top wall and a non-reflective outflow on the right.

- `modmesh/multidim/euler/oblique.py`: `ObliqueShockMesher` takes the domain geometry (`nx`, `ny`, `ll`, `ur`, `x_ramp`, `wedge_angle`) and `make_mesh(cell_type)` builds the domain in three element flavors: `'quad'` (one quadrilateral per grid box), `'triangle'` (each box split along its lower-left-to-upper-right diagonal), and `'unstructured'` (Bowyer-Watson Delaunay triangulation of the boundary nodes plus deterministically jittered interior points).
  - Every flavor collects its nodes in a `PointPadFp64` through the `SimpleArray` API.
  - In either triangular flavor, No cell touches the boundary with more than one face.
  - `classify_boundary` buckets the boundary faces into inlet, slip-wall, and outflow lists ready for the boundary-condition registration.
- `modmesh/pilot/_oblique.py`: pilot menu items drawing the three flavors.
- `tests/test_multidim_euler_oblique.py`: mesh shape and combinatorics, boundary classification, cell winding, domain tiling, determinism, and the one-boundary-face-per-cell rule.

For phase 1 of issue #404.

### CODE REVIEW

File: modmesh/multidim/euler/oblique.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+import math
+
+from ... import core
+
+__all__ = [
+    'ObliqueShockMesher',
+]
+
+
+class ObliqueShockMesher(object):
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:49]: This is brand-new code for generating quadrilateral and triangular meshes.

File: modmesh/multidim/euler/oblique.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        self.x_ramp = x_ramp
+        self.tan_theta = math.tan(math.radians(wedge_angle))
+        wedge_top = (self.x1 - x_ramp) * self.tan_theta
+        if wedge_top >= self.y1 - self.y0:
+            raise ValueError(
+                f"wedge top (ur[0] - x_ramp) * tan(wedge_angle) = "
+                f"{wedge_top:g} must stay below the domain height "
+                f"{self.y1 - self.y0:g}")
+
+    def _node(self, it, jt):
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:74]: `_node()` is a helper to determine the coordinates for the wedged domain.

File: modmesh/multidim/euler/oblique.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        return x, yb + (self.y1 - yb) * jt / self.ny
+
+    def _nid(self, it, jt):
+        return jt * (self.nx + 1) + it
+
+    def _box(self, it, jt):
+        return (self._nid(it, jt), self._nid(it + 1, jt),
+                self._nid(it + 1, jt + 1), self._nid(it, jt + 1))
+
+    def make_mesh(self, cell_type='quad'):
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:87]: Return a `StaticMesh` based on the input `cell_type`.

File: modmesh/multidim/euler/oblique.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        Both triangular flavors keep at most one boundary face per cell,
+        which the corner ``'quad'`` cells cannot.  All flavors share the
+        boundary layout (nx segments on the bottom and top, ny on the left
+        and right) and produce counter-clockwise cells.  The returned mesh
+        has ``ndcrd``/``cltpn``/``clnds`` filled and ``build_interior`` /
+        ``build_boundary`` / ``build_ghost`` run.
+        """
+        nx, ny = self.nx, self.ny
+        if cell_type in ('quad', 'triangle'):
+            nodes = core.PointPadFp64(ndim=2)
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:108]: Use `PointPadFp64` to hold coordinate data.

File: modmesh/multidim/euler/oblique.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+                             ncell=len(cells))
+        mh.ndcrd[:, :] = nodes.pack_array().ndarray
+        mh.cltpn.fill(tpn)
+        mh.clnds[:, :len(cells[0])] = cells
+        mh.build_interior(do_metric=True)
+        mh.build_boundary()
+        mh.build_ghost()
+        return mh
+
+    def _jitter_points(self):
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:147]: Prototyped Delaunay triangular mesh generation.

File: modmesh/multidim/euler/oblique.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        ``pad``; returns CCW index triples.
+
+        Seed a far-away super-triangle, insert one point at a time (carve
+        the triangles whose circumcircle strictly contains it, fan the
+        cavity rim to the point), then drop the triangles touching a super
+        vertex.  On-circle points count as outside, keeping the cavity well
+        defined under the cocircular ties of the regular boundary layout.
+        Naive O(n^2).
+
+        References:
```
Comment (yungyuc) [inline modmesh/multidim/euler/oblique.py:197]: Reference to the Delaunay triangulation.

File: modmesh/pilot/_oblique.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        )
+        self._add_menu_item(
+            menu=self._mgr.meshMenu,
+            text="Sample: oblique-shock reflection mesh (2D unstructured)",
+            tip="Draw the unstructured (Delaunay) triangle wedge mesh for "
+                "the oblique-shock reflection",
+            func=self.draw_unstructured_mesh,
+        )
+
+    def draw_quad_mesh(self):
```
Comment (yungyuc) [inline modmesh/pilot/_oblique.py:72]: Add 3 windows to draw the mesh.  They will be refactored for a better GUI in the future.

File: tests/test_multidim_euler_oblique.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    NY = 8
+    LL = (0.0, 0.0)
+    UR = (3.0, 1.0)
+    CELL_TYPE = None
+    CELLS_PER_BOX = None
+    CLTPN = None
+
+    @classmethod
+    def setUpClass(cls):
+        cls.mesher = ObliqueShockMesher(nx=cls.NX, ny=cls.NY, ll=cls.LL,
```
Comment (yungyuc) [inline tests/test_multidim_euler_oblique.py:67]: Make mesh.

File: tests/test_multidim_euler_oblique.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        mh = self.mesh
+        per_cell = {}
+        for ifc in self._boundary_faces():
+            icl = mh.fccls[ifc, 0]
+            per_cell[icl] = per_cell.get(icl, 0) + 1
+        self.assertTrue(per_cell)
+        self.assertLessEqual(max(per_cell.values()), 1)
+
+
+class ObliqueShockQuadMeshTC(_ObliqueMeshBase, unittest.TestCase):
```
Comment (yungyuc) [inline tests/test_multidim_euler_oblique.py:204]: Test for quadrilateral mesh.

File: tests/test_multidim_euler_oblique.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+    """One quadrilateral per grid box."""
+
+    CELL_TYPE = 'quad'
+    CELLS_PER_BOX = 1
+    CLTPN = modmesh.StaticMesh.QUADRILATERAL
+
+
+class ObliqueShockTriangleMeshTC(_ObliqueMeshBase, _SingleBoundaryFaceTB,
```
Comment (yungyuc) [inline tests/test_multidim_euler_oblique.py:212]: Test for regular triangular mesh.

File: tests/test_multidim_euler_oblique.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+                                 unittest.TestCase):
+    """Each grid box cut along its diagonal into two triangles."""
+
+    CELL_TYPE = 'triangle'
+    CELLS_PER_BOX = 2
+    CLTPN = modmesh.StaticMesh.TRIANGLE
+
+
+class ObliqueShockUnstructuredMeshTC(_ObliqueMeshBase, _SingleBoundaryFaceTB,
```
Comment (yungyuc) [inline tests/test_multidim_euler_oblique.py:221]: Test for unstructured triangular mesh.
